proxy_ring_buffer.go

```diff
@@ -0,0 +1,278 @@
+// to find that item.
+//
+// Add returns the sequence number of the added item. You can read the added item using this number.
+func (rb *Ringbuffer) Add(ctx context.Context, item interface{}, overflowPolicy OverflowPolicy) (sequence int64, err error) {
+	elementData, err := rb.validateAndSerialize(item)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	request := codec.EncodeRingbufferAddRequest(rb.name, int32(overflowPolicy), elementData)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferAddResponse(response), nil
+}
+
+// AddAll
+// Adds all the items of a collection to the tail of the Ringbuffer. A addAll is likely to outperform multiple calls
+// to add(Object) due to better io utilization and a reduced number of executed operations. If the batch is empty,
+// the call is ignored. When the collection is not empty, the content is copied into a different data-structure.
+// This means that: after this call completes, the collection can be re-used. the collection doesn't need to be serializable.
+// If the collection is larger than the capacity of the Ringbuffer, then the items that were written first will be
+// overwritten. Therefor this call will not block. The items are inserted in the order of the Iterator of the collection.
+// If an addAll is executed concurrently with an add or addAll, no guarantee is given that items are contiguous.
+// The result of the future contains the sequenceId of the last written item.
+func (rb *Ringbuffer) AddAll(ctx context.Context, overflowPolicy OverflowPolicy, items ...interface{}) (int64, error) {
+	elementData, err := rb.validateAndSerializeValues(items)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	request := codec.EncodeRingbufferAddAllRequest(rb.name, elementData, int32(overflowPolicy))
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
+	}
+	return codec.DecodeRingbufferAddAllResponse(response), nil
+}
+
+// ReadOne
+// Reads one item from the Ringbuffer. If the sequence is one beyond the current tail, this call blocks until an
+// item is added. This method is not destructive unlike e.g. a queue.take. So the same item can be read by multiple
+// readers, or it can be read multiple times by the same reader. Currently, it isn't possible to control how long this
+// call is going to block. In the future we could add e.g. tryReadOne(long sequence, long timeout, TimeUnit unit).
+func (rb *Ringbuffer) ReadOne(ctx context.Context, sequence int64) (interface{}, error) {
+	request := codec.EncodeRingbufferReadOneRequest(rb.name, sequence)
+	response, err := rb.invokeOnPartition(ctx, request, rb.partitionID)
+	if err != nil {
+		return ReadResultSetSequenceUnavailable, err
```

Comment:
How about `return nil, err` on this line? Same comment for other lines which have `return ReadResultSetSequenceUnavailable, err`

Reply:
For the ReadOne it makes sense and I changed it - see 172728c477ae3d0b93505108dd130f01c8afeef5
For other places, where return type is int64, one can't use nil, but must be ReadResultSetSequenceUnavailable instead